cpp_const: Add ability to impose C++ const semantics in Python.

EricCousineau-TRI opened the pull request:
Copying from #7841:

> This implements functionality to enable a potential solution for #7793, and to provide workarounds for https://github.com/pybind/pybind11/issues/1241 (callbacks and virtual function overrides - #7838).

C++ Pybind Code - [cpp_const_pybind_test_py.cc](https://github.com/EricCousineau-TRI/repro/blob/1cb311b/python/bindings/pymodule/const/test/cpp_const_pybind_test_py.cc)
Python Code, using C++ - [cpp_const_pybind_test.py](https://github.com/EricCousineau-TRI/repro/blob/1cb311b/python/bindings/pymodule/const/test/cpp_const_pybind_test.py)

This imports a portion of [`wrapt`](https://github.com/GrahamDumpleton/wrapt/tree/e0d82df6dbdd3d602896fc528942ffa1902e2d5e) with slight modifications (shown in the commits), and enables the pure-Python portion of the feature.

Please note that this PR exceeds the nominal line count due to third-party code; however, excluding the third party code, the line count is 421 insertions:
```
$ git diff --shortstat HEAD~ -- bindings/
 3 files changed, 421 insertions(+)
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7884)
<!-- Reviewable:end -->

EricCousineau-TRI commented:
+@soonho-tri for feature review, if you have time, please.
+@jwnimmer-tri for platform review on Monday, please. (Let me know if you want this code incorporated in a different fashion.)

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-:-L3pa14iBU6S6Kt8EWnl:b-8qhkx0)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
> (Let me know if you want this code incorporated in a different fashion.)

You mean the `//third_party/...`?  Seems fine to vendor it, since you want to make edits.  It looks like our pydrake install rule installs pydrake code that depends on it though, so you'll need to make sure the installed bindings work, and that the new LICENSE file gets installed.

soonho-tri commented:
Reviewed 4 of 5 files at r1, 2 of 4 files at r2.
Review status: 6 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 5 files at r1, 2 of 4 files at r2.
Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41c9O72qDrf4HHYkG7:-L41c9O72qDrf4HHYkG8:bh4h2q3) (no related file):*
BTW It would probably be cleaner if the first commit (importing from upstream) had only the files actually from upstream + the README; as of now it also has the init and BUILD files.  A simple "git blame" ends up being misleading right now, though if someone consulted the README they could figure it out.

---

*[bindings/pydrake/util/BUILD.bazel, line 127 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41jSN-2EAtD-_APLmF:-L41jSN0YUQACF8_YhND:b-kyn78s) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/BUILD.bazel#L127)):*
> ```Python
> UPSTREAM_INSTALLS = [
>     "//third_party/wrapt:install",
> ]
> ```

If I recall correctly, our convention is to have something more like `//tools/workspace:BUILD.bazel` handle the third-party license installs.  Another option would be to have `//third_party:BUILD.bazel` have the install rule.  In any case, burying it here is surprising enough to be defective.

---

*[third_party/wrapt/BUILD.bazel, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41jC4bnfQVj9QeRdea:-L41jC4bnfQVj9QeRdeb:b-a9hklf) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/third_party/wrapt/BUILD.bazel#L23)):*
> ```Python
>     name = "install",
>     targets = [":wrapt"],
>     py_dest = "lib/python2.7/dist-packages/wrapt",
> ```

TBD Does this go into the right place?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[third_party/wrapt/BUILD.bazel, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41jC4bnfQVj9QeRdea:-L41uB_uNgGv0Z41h2pu:b-khhj5b) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/third_party/wrapt/BUILD.bazel#L23)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

TBD Does this go into the right place?
</blockquote></details>

(Sorry, this was a note to myself.  I'll follow-up later.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Per f2f with @jwnimmer-tri, we looked into using Ubuntu Debian packages / Mac PIP packages. However, the Ubuntu package is tagged with `1.8.0`, and looking at the diff of `wrapt` between `1.8.0` and `1.10.11`, it looks like there was bug in which `__mod__` was implemented using `a ^ b` rather than `a % b`.

Given this, will use a fork for the time being. I will add a TODO to drop the fork once later versions of `python-wrapt` are available on Ubuntu.

---

Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[third_party/wrapt/BUILD.bazel, line 23 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41jC4bnfQVj9QeRdea:-L41tUHfCv9WtKaeYfiQ:b-2pwb6y) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/third_party/wrapt/BUILD.bazel#L23)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

TBD Does this go into the right place?
</blockquote></details>

OK Actually, I'm realizing that I want this to go to `drake.third_party.wrapt` - I don't want our changes to conflict with other people's usages.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-:-L41zBAIEkwIZNeWSXTn:b4kmk45)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 4 files at r2.
Review status: 7 of 8 files reviewed at latest revision, 19 unresolved discussions.

---

*[bindings/pydrake/util/cpp_const.py, line 2 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41tvu0XyLa66Wp4WMQ:-L41tvu0XyLa66Wp4WMR:bzbw2ha) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L2)):*
> ```Python
> """
> @file
> ```

BTW Is `@file` a thing python community and/or pydoc knows about?  I would have thought that the top level `__doc__` was already understood to be file-wide overview?

---

*[bindings/pydrake/util/cpp_const.py, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41um9gCWeLAweSKaNj:-L41um9hqCBLeeckhjmd:b-l3fvun) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L11)):*
> ```Python
> 
> # N.B. There should *always* be an option for disabling this for performance
> # reasons!
> ```

I am not sure what I, as a reader, am supposed to take away from reading this comment.

---

*[bindings/pydrake/util/cpp_const.py, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42BIsWUbWpuMNAhIG9:-L42BIsWUbWpuMNAhIGA:b-m7068z) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L18)):*
> ```Python
> 
> 
> class ConstError(TypeError):
> ```

Probably requires pydoc class overview per styleguide?

---

*[bindings/pydrake/util/cpp_const.py, line 27 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42GQNEkxtwwacW1aCN:-L42GQNF00Jk6qhwW_TR:b8jp68d) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L27)):*
> ```Python
>         self._cls = cls
>         self._owned_properties = set(owned_properties or set())
>         self.mutable_methods = set(mutable_methods or set())
> ```

FYI Access to this isn't called in a tight loop, so consider hiding it behind an accessor for privacy?  Alternatively -- unclear why some members are private and some are public.

---

*[bindings/pydrake/util/cpp_const.py, line 50 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42GC6Je4fV16_12fce:-L42GC6Je4fV16_12fcf:bxtib36) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L50)):*
> ```Python
>         # Limitation: This would not handle overloads (e.g.
>         # `const T& get() const` and `T& get()`). However, C++ should decorate
>         # the methods such that we should not need to mark them.
> ```

Unclear what `decorate` means here.  What specific mechanism?

---

*[bindings/pydrake/util/cpp_const.py, line 82 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42Fjhz2EOKh1RE1EsR:-L42Fji-q_e9nhFWSkIn:b-tn74wa) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L82)):*
> ```Python
> _const_metas = _ConstClassMetaMap()
> # Register common mutators.
> # N.B. These methods actually have to be overridde in `Const`, since neither
> ```

BTW Typo `ridde`n.

---

*[bindings/pydrake/util/cpp_const.py, line 98 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42Hcb0CCRgvqjDuNUN:-L42Hcb0CCRgvqjDuNUO:b-sa7y2) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L98)):*
> ```Python
>     "__imod__",
>     "__ipow__",
>     "__ilshift__",
> ```

Missing `__irshift__`.

---

*[bindings/pydrake/util/cpp_const.py, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42HzpvStPtAqNzZJBn:-L42HzpvStPtAqNzZJBo:buk1y27) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L111)):*
> ```Python
>     'insert',
>     'pop',
>     'remove',
> ```

Missing `reverse`.

---

*[bindings/pydrake/util/cpp_const.py, line 117 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42ICT_rn4jGuE5L4YF:-L42ICT_rn4jGuE5L4YG:b-hr5jb3) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L117)):*
> ```Python
>     'clear',
>     'setdefault',
>     })
> ```

Missing:
`__missing__`
`pop`
`popitem`
... of if they are already handled (because they delegate to another method) please comment as such.

---

*[bindings/pydrake/util/cpp_const.py, line 150 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42BvxXCf9xhjS-wAx9:-L42BvxXCf9xhjS-wAxA:b-ii0pkg) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L150)):*
> ```Python
>     def __str__(self):
>         out = self.__wrapped__.__str__()
>         if (len(out) >= 2 and len(out) < 200 and
> ```

I couldn't figure out the reason for the 200 limit.

---

*[bindings/pydrake/util/cpp_const.py, line 159 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42D7kq0GqEeBgq6d-O:-L42D7kq0GqEeBgq6d-P:b-6q5hdd) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L159)):*
> ```Python
> def _install_mutable_error_methods():
>     # Automatically rewrite any mutable methods for `object` to deny access
>     # upon calling.
> ```

It took me a while before I realized that this was rebinding `_Const`'s inherited methods from `object` -- specifically, that "any mutable methods for `object`" meant "on the `_Const` class".  (I thought we were intercepting all object methods everywhere!)

---

*[bindings/pydrake/util/cpp_const.py, line 194 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41vy3xFqeQbY26-s6x:-L41vy3xFqeQbY26-s6y:bldyvc2) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L194)):*
> ```Python
> def to_const(obj):
>     """Converts an object to a const proxy. Does not proxy immutable / literal
>     types. """
> ```

BTW _Meta_ No whitespace between `. "`, throughout.

---

*[bindings/pydrake/util/cpp_const.py, line 194 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41wjXeySklAXXcrK-4:-L41wjXeySklAXXcrK-5:b8h8cl9) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L194)):*
> ```Python
> def to_const(obj):
>     """Converts an object to a const proxy. Does not proxy immutable / literal
>     types. """
> ```

FYI I assume basically _any_ kind of object could be passed in here?  If yes, great.  If not and there's things I shouldn't pass in here, then please document then.

---

*[bindings/pydrake/util/cpp_const.py, line 232 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41wETtVgJjyQ6LQ8ni:-L41wETtVgJjyQ6LQ8nj:b-fkjn7y) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L232)):*
> ```Python
>     direct type.
>     """
>     if isinstance(obj, _Const):
> ```

Rule-of-3 duplication of this check (just calling the existing helper?).

---

*[bindings/pydrake/util/cpp_const.py, line 244 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41wUylgnVVZiUGWGOO:-L41wUylgnVVZiUGWGOP:blyi4u9) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L244)):*
> ```Python
> 
> 
> def mutable_method(func):
> ```

`mutable_method` is a bit of a confusing name, because its missing a verb.

Ditto for `const_meta` below.

---

*[bindings/pydrake/util/cpp_const.py, line 257 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42AntV_1xUMr18rucX:-L42AntV_1xUMr18rucY:b-z58d05) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L257)):*
> ```Python
> 
> def const_meta(owned_properties=None, mutable_methods=None):
>     """Decorates a class with const-proxy metadata. """
> ```

Actually, it returns a functor that does so?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 5 files at r2, 1 of 16 files at r3.
Review status: 2 of 12 files reviewed at latest revision, 21 unresolved discussions, some commit checks failed.

---

*[bindings/pydrake/util/cpp_const.py, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42Z0jzvyHAB_wWvQza:-L42Z0k-J_bghCORF2T1:bx9ees4) ([raw file](https://github.com/robotlocomotion/drake/blob/007a0c9236f36e8a7550f8532d3674a4fc179c0e/bindings/pydrake/util/cpp_const.py#L26)):*
> ```Python
>     def __init__(self, cls, owned_properties=None, mutable_methods=None):
>         self._cls = cls
>         self._owned_properties = set(owned_properties or set())
> ```

FYI Should it be an error here if `owned_properties` mentions a field that doesn't exist on `cls`?  Otherwise it seems to me like typos in using-code might go unnoticed?

---

*[bindings/pydrake/util/test/cpp_const_test.py, line 4 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42YEfjMUekO_ld2R59:-L42YEfjMUekO_ld2R5A:b-w843ni) ([raw file](https://github.com/robotlocomotion/drake/blob/007a0c9236f36e8a7550f8532d3674a4fc179c0e/bindings/pydrake/util/test/cpp_const_test.py#L4)):*
> ```Python
> import unittest
> 
> import pydrake.util.cpp_const as m
> ```

FYI I am not loving `m` as a global variable name; usually 1-letter variables would be loop iterators, or at least locals.  I think its borderline enough that I'm not going to hard defect it, but consider sticking with `cpp_const` for clarity, or perhaps `as mut  # Module under test.` to split the difference on line wrap vs readability.  (Or is `m` a pythonic habit that I just haven't seen yet?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 12 files reviewed at latest revision, 21 unresolved discussions.

---

*[bindings/pydrake/util/BUILD.bazel, line 127 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41jSN-2EAtD-_APLmF:-L42OzBfSD6oFlTdugV7:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/BUILD.bazel#L127)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If I recall correctly, our convention is to have something more like `//tools/workspace:BUILD.bazel` handle the third-party license installs.  Another option would be to have `//third_party:BUILD.bazel` have the install rule.  In any case, burying it here is surprising enough to be defective.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 2 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41tvu0XyLa66Wp4WMQ:-L42jDLeUR1_J4sdYSMR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L2)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Is `@file` a thing python community and/or pydoc knows about?  I would have thought that the top level `__doc__` was already understood to be file-wide overview?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 11 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41um9gCWeLAweSKaNj:-L42fwuiCVs81hNqMxEb:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L11)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I am not sure what I, as a reader, am supposed to take away from reading this comment.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 27 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42GQNEkxtwwacW1aCN:-L42eIABmj3-Jpcm0vy-:b6cuh6n) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L27)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Access to this isn't called in a tight loop, so consider hiding it behind an accessor for privacy?  Alternatively -- unclear why some members are private and some are public.
</blockquote></details>

OK `mutable_methods` is public because it is used by `_install_object_mutable_methods`, but it's the only property used.

---

*[bindings/pydrake/util/cpp_const.py, line 50 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42GC6Je4fV16_12fce:-L42eFyxYCyud8sx6TzE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L50)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear what `decorate` means here.  What specific mechanism?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 82 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42Fjhz2EOKh1RE1EsR:-L42eFPYkJaSpG4ZuP76:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L82)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Typo `ridde`n.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 98 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42Hcb0CCRgvqjDuNUN:-L42eEgMmt0cS2pwmcX9:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L98)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Missing `__irshift__`.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 111 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42HzpvStPtAqNzZJBn:-L42eEQq1m4ntYqwhBfC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L111)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Missing `reverse`.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 117 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42ICT_rn4jGuE5L4YF:-L42e1T4VXQh8DlRycyJ:b-83sxqx) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L117)):*
Done.

`__missing__` is not intrinsically mutable:
```
class Dict(dict):
    def __missing__(self, key):
        return key
...
>>> d = Dict(a=1)
>>> d["b"]
'b'
>>> d
{'a': 1}
>>> 
```

---

*[bindings/pydrake/util/cpp_const.py, line 150 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42BvxXCf9xhjS-wAx9:-L42g-bJBdCDdjFbWglT:bgmifus) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L150)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I couldn't figure out the reason for the 200 limit.
</blockquote></details>

Done. Removed limit, and ensured that this only applies to `repr`.

---

*[bindings/pydrake/util/cpp_const.py, line 159 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42D7kq0GqEeBgq6d-O:-L42g2gKwaARJntwaZ2u:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L159)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

It took me a while before I realized that this was rebinding `_Const`'s inherited methods from `object` -- specifically, that "any mutable methods for `object`" meant "on the `_Const` class".  (I thought we were intercepting all object methods everywhere!)
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 194 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41wjXeySklAXXcrK-4:-L42ibIYnVTXSmJIqCHm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L194)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I assume basically _any_ kind of object could be passed in here?  If yes, great.  If not and there's things I shouldn't pass in here, then please document then.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 194 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41vy3xFqeQbY26-s6x:-L42ibUnaIntVIBYBYxS:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L194)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW _Meta_ No whitespace between `. "`, throughout.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 232 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41wETtVgJjyQ6LQ8ni:-L42iaQoe0Czth6GBtKp:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L232)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Rule-of-3 duplication of this check (just calling the existing helper?).
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 244 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41wUylgnVVZiUGWGOO:-L42RI_LPi1JJFrqTPS6:b7salmr) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L244)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

`mutable_method` is a bit of a confusing name, because its missing a verb.

Ditto for `const_meta` below.
</blockquote></details>

OK I'd prefer to keep it in-sync with other Python decorators, like [`staticmethod`](https://docs.python.org/2/library/functions.html#staticmethod) or [`property`](https://docs.python.org/2/library/functions.html#property).

Changed `const_meta` to `const_decorated`, though, as `const_class` nor `mutable_class` make much of any sense.

---

*[bindings/pydrake/util/cpp_const.py, line 257 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42AntV_1xUMr18rucX:-L42RenZF4kSFqrrWYbJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L257)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Actually, it returns a functor that does so?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42Z0jzvyHAB_wWvQza:-L42fYc3sm-o7R9_nAAu:b-w0mrlj) ([raw file](https://github.com/robotlocomotion/drake/blob/007a0c9236f36e8a7550f8532d3674a4fc179c0e/bindings/pydrake/util/cpp_const.py#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Should it be an error here if `owned_properties` mentions a field that doesn't exist on `cls`?  Otherwise it seems to me like typos in using-code might go unnoticed?
</blockquote></details>

OK I'd prefer not to restrict this. Most classes that I've seen do not indicate their members until `__init__` is called, and the properties are added dynamically (and not really as properties per se, but entries in the instance's `__dict__`.

---

*[bindings/pydrake/util/test/cpp_const_test.py, line 4 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42YEfjMUekO_ld2R59:-L42jaXgGn5B-6EVO60K:b-t9y5zb) ([raw file](https://github.com/robotlocomotion/drake/blob/007a0c9236f36e8a7550f8532d3674a4fc179c0e/bindings/pydrake/util/test/cpp_const_test.py#L4)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I am not loving `m` as a global variable name; usually 1-letter variables would be loop iterators, or at least locals.  I think its borderline enough that I'm not going to hard defect it, but consider sticking with `cpp_const` for clarity, or perhaps `as mut  # Module under test.` to split the difference on line wrap vs readability.  (Or is `m` a pythonic habit that I just haven't seen yet?)
</blockquote></details>

Done. It's a convention used in `pybind11`, but changed it to `cpp_const` for clarity.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 12 files reviewed at latest revision, 14 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41c9O72qDrf4HHYkG7:-L42jujV3Z7GEGAgj1Hv:b-896fix) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW It would probably be cleaner if the first commit (importing from upstream) had only the files actually from upstream + the README; as of now it also has the init and BUILD files.  A simple "git blame" ends up being misleading right now, though if someone consulted the README they could figure it out.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 0 of 12 files reviewed at latest revision, 13 unresolved discussions.

---

*[bindings/pydrake/util/cpp_const.py, line 18 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42BIsWUbWpuMNAhIG9:-L42jxsn5rZ7bAKsYd8j:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L18)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Probably requires pydoc class overview per styleguide?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 13 of 16 files at r3, 2 of 2 files at r4.
Review status: 10 of 12 files reviewed at latest revision, 4 unresolved discussions.

---

*[bindings/pydrake/util/cpp_const.py, line 117 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42ICT_rn4jGuE5L4YF:-L42nG_g14DFjodiLFw-:b-xw6jvo) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L117)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done.

`__missing__` is not intrinsically mutable:
```
class Dict(dict):
    def __missing__(self, key):
        return key
...
>>> d = Dict(a=1)
>>> d["b"]
'b'
>>> d
{'a': 1}
>>> 
```
</blockquote></details>

OK Ah, yes!  I misread the python docs.

---

*[bindings/pydrake/util/cpp_const.py, line 244 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41wUylgnVVZiUGWGOO:-L42oZLfWA5_ChVM-3zT:b3io6c4) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L244)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

OK I'd prefer to keep it in-sync with other Python decorators, like [`staticmethod`](https://docs.python.org/2/library/functions.html#staticmethod) or [`property`](https://docs.python.org/2/library/functions.html#property).

Changed `const_meta` to `const_decorated`, though, as `const_class` nor `mutable_class` make much of any sense.
</blockquote></details>

Aha, I didn't realize (while reading) that these are intended to be used as Python decorators.  In that case, the names make perfect sense, but the documentation is unclear.

---

*[bindings/pydrake/util/cpp_const.py, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42Z0jzvyHAB_wWvQza:-L42mg5pAd9PTqO5Wyby:b-h5vwe6) ([raw file](https://github.com/robotlocomotion/drake/blob/007a0c9236f36e8a7550f8532d3674a4fc179c0e/bindings/pydrake/util/cpp_const.py#L26)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

OK I'd prefer not to restrict this. Most classes that I've seen do not indicate their members until `__init__` is called, and the properties are added dynamically (and not really as properties per se, but entries in the instance's `__dict__`.
</blockquote></details>

OK Good point.

---

*[bindings/pydrake/util/cpp_const.py, line 232 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42o6YWEioQ0klYJIOr:-L42o6YWEioQ0klYJIOs:b-3g5ci0) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L232)):*
> ```Python
>     WARNING: Do NOT use this for branching in production code unless
>     const-proxying is guaranteed to be enabled or the branching is designed
>     not to fail in this case.
> ```

I guess we can't name it `_is_const` because we want to unit test it?

The warning seems dire enough (and correctly so) that the method name should perhaps reflect that?  Or we could promise to fail-fast here if const checking got compiled out or disabled?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
LGTM Modulo future changes from @soonho-tri review.

---

Reviewed 2 of 16 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[bindings/pydrake/BUILD.bazel, line 140 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42qDU2L9WC6zED-bQC:-L42qDU2L9WC6zED-bQD:bhl9v08) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/BUILD.bazel#L140)):*
> ```Python
>     "//drake/bindings/pydrake/third_party",
>     "//drake/bindings/pydrake/util",
> ])
> ```

FYI Is the name `PYBIND_LIBRARIES` still applicable, with this new item added?

---

*[bindings/pydrake/third_party/forward_files.bzl, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42r6_juA8XjmqZMXA7:-L42r6_k3UCJxt4OHrad:b-1xzd7y) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/third_party/forward_files.bzl#L6)):*
> ```Python
>         srcs,
>         strip_prefix,
>         dest_prefix,
> ```

BTW I believe its skylark style to use kwargs syntax (`= None`) for all non-`name` arguments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-:-L42rEKuWwv0N1ius1ZF:b-lekgo9)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42sqGq3RpA7ozIAA9P:-L42sqGq3RpA7ozIAA9Q:b-eu29ef) (no related file):*
FYI I've confirmed that the installed files end up in good places ...
```
lib/python2.7/site-packages/pydrake/third_party/wrapt
lib/python2.7/site-packages/pydrake/third_party/wrapt/__init__.py
lib/python2.7/site-packages/pydrake/third_party/wrapt/wrappers.py
share/doc/wrapt
share/doc/wrapt/LICENSE
```
... and that I can `import pydrake.util.cpp_const` from the installed image.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Checkpoint.

---

Reviewed 15 of 16 files at r3, 1 of 2 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[bindings/pydrake/util/cpp_const.py, line 3 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42U8psGcfKLhH13bst:-L42U8psGcfKLhH13bsu:b-sf758t) ([raw file](https://github.com/robotlocomotion/drake/blob/007a0c9236f36e8a7550f8532d3674a4fc179c0e/bindings/pydrake/util/cpp_const.py#L3)):*
> ```Python
> """
> @file
> Enables a semi-transparent layer of C++ const-honoring Python proxies.
> ```

Can you explain why this is "semi-" transparent?

---

*[bindings/pydrake/util/cpp_const.py, line 41 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42qRLAIeJ5lzdJTx7B:-L42qRLAIeJ5lzdJTx7C:bapfecw) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L41)):*
> ```Python
> 
>     def is_owned_property(self, name):
>         # Determines if a property is owned, and if the returned value should
> ```

Is "a property is owned" the same as "the returned value should be `const`"?

---

*[bindings/pydrake/util/cpp_const.py, line 85 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42tKDxGGLqIajo5k_D:-L42tKDxGGLqIajo5k_E:b-eyo1wa) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L85)):*
> ```Python
> # N.B. All methods registered will respect inheritance.
> # N.B. For `object`, see `_install_object_mutable_methods`.
> _const_metas.emplace(object, mutable_methods={
> ```

Is it a list of built-in methods available in Python `object`? As a reviewer, I'd like to check if this list is sound and complete but I do not have a clue how to do that (i.e. `dir(object)` gives a different lists, not a super set of it, for instance). Ditto for `list` and `dict` below.

---

*[bindings/pydrake/util/cpp_const.py, line 183 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42z0n-AWlz0VYnV1LY:-L42z0n-AWlz0VYnV1LZ:b-33vg4u) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L183)):*
> ```Python
>     # Detects if a type is a immutable (or literal) type.
>     literal_types = [int, float, str, unicode, tuple, type(None)]
>     if type(obj) in literal_types:
> ```

BTW, this can be just `return type(obj) in literal_types`?

---

*[bindings/pydrake/util/cpp_const.py, line 232 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L43-VtAzOz7-G_KtqcD:-L43-VtAzOz7-G_KtqcE:b151uuc) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L232)):*
> ```Python
>     WARNING: Do NOT use this for branching in production code unless
>     const-proxying is guaranteed to be enabled or the branching is designed
>     not to fail in this case.
> ```

FYI, I don't understand "the branching is designed not to fail in this case" part. Could you elaborate?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-:-L4301nQSKDiViDz2hdD:by96gdd)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 7 of 17 files reviewed at latest revision, 10 unresolved discussions.

---

*[bindings/pydrake/BUILD.bazel, line 140 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42qDU2L9WC6zED-bQC:-L43tmMAPL6Sf1dEgG2i:b5rpljv) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/BUILD.bazel#L140)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Is the name `PYBIND_LIBRARIES` still applicable, with this new item added?
</blockquote></details>

Done. Made a rename across `pydrake`, can revert that commit if it's scope creep.

---

*[bindings/pydrake/third_party/forward_files.bzl, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42r6_juA8XjmqZMXA7:-L43rlDq4gmLLFXr2dpr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/third_party/forward_files.bzl#L6)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I believe its skylark style to use kwargs syntax (`= None`) for all non-`name` arguments.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 244 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L41wUylgnVVZiUGWGOO:-L43s1YO2u7IUpKI24xa:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/aa85bbcb2b8fddedc778c7e669ca63c8db989e36/bindings/pydrake/util/cpp_const.py#L244)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Aha, I didn't realize (while reading) that these are intended to be used as Python decorators.  In that case, the names make perfect sense, but the documentation is unclear.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 3 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42U8psGcfKLhH13bst:-L43rm-FgE5BR1aAiidF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/007a0c9236f36e8a7550f8532d3674a4fc179c0e/bindings/pydrake/util/cpp_const.py#L3)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Can you explain why this is "semi-" transparent?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 41 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42qRLAIeJ5lzdJTx7B:-L43rmq9l3gCYHIN_tXh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L41)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Is "a property is owned" the same as "the returned value should be `const`"?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 85 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42tKDxGGLqIajo5k_D:-L43rnj4UyPHdCweBvbH:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L85)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Is it a list of built-in methods available in Python `object`? As a reviewer, I'd like to check if this list is sound and complete but I do not have a clue how to do that (i.e. `dir(object)` gives a different lists, not a super set of it, for instance). Ditto for `list` and `dict` below.
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 183 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42z0n-AWlz0VYnV1LY:-L43roXbeOutsh5wI_sC:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L183)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, this can be just `return type(obj) in literal_types`?
</blockquote></details>

Done.

---

*[bindings/pydrake/util/cpp_const.py, line 232 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42o6YWEioQ0klYJIOr:-L43rqERojDvN-SHxRFx:b-d5zrot) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L232)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I guess we can't name it `_is_const` because we want to unit test it?

The warning seems dire enough (and correctly so) that the method name should perhaps reflect that?  Or we could promise to fail-fast here if const checking got compiled out or disabled?
</blockquote></details>

Done.
Yuppers, and I'd prefer not to make the name confined to just testing, as it is used in `to_const` and `to_mutable`. I've added the `*_test` suffix to (faintly) try to make people scratch their heads.

---

*[bindings/pydrake/util/cpp_const.py, line 232 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L43-VtAzOz7-G_KtqcD:-L43s0WQWuJ4RVjjrGVA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/util/cpp_const.py#L232)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

FYI, I don't understand "the branching is designed not to fail in this case" part. Could you elaborate?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[bindings/pydrake/third_party/forward_files.bzl, line 6 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L42r6_juA8XjmqZMXA7:-L448H2vl2NlT46fcf1n:b6ixq9h) ([raw file](https://github.com/robotlocomotion/drake/blob/5bfad7439abe784aaa737ad9aeb939ea6c8590ab/bindings/pydrake/third_party/forward_files.bzl#L6)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done.
</blockquote></details>

BTW also `srcs`?

---

*[tools/skylark/pybind.bzl, line 156 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L448nwDCmtwQyb7l39U:-L448nwDCmtwQyb7l39V:b-9ntieu) ([raw file](https://github.com/robotlocomotion/drake/blob/c0e71feb2b272bff5400a73f04a3815430738951/tools/skylark/pybind.bzl#L156)):*
> ```Python
> 
> def _get_install(target):
>     # Gets the install target for a Python target that have a sibling install
> ```

BTW typo `has`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-:-L449Lwqt01IBnhk_yuK:bnfp4nl)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 10 of 10 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/util/cpp_const.py, line 166 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L46BW_m2jCbGpacZRSA:-L46BW_m2jCbGpacZRSB:b5sip7c) ([raw file](https://github.com/robotlocomotion/drake/blob/c0e71feb2b272bff5400a73f04a3815430738951/bindings/pydrake/util/cpp_const.py#L166)):*
> ```Python
> 
>     def __dict_custom__(self):
>         # Prevent direct mutation of dictionary.
> ```

BTW, I was wondering if it's safe to use double-underscore (`__`) prefix for a non-system method such as this one. (This is really a question for my Python education).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-:-L46C560GAgo3Se6diDl:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 14 of 17 files reviewed at latest revision, 2 unresolved discussions.

---

*[bindings/pydrake/util/cpp_const.py, line 166 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L46BW_m2jCbGpacZRSA:-L46tOlWvxKxmhLsJdPI:b-p20a5m) ([raw file](https://github.com/robotlocomotion/drake/blob/c0e71feb2b272bff5400a73f04a3815430738951/bindings/pydrake/util/cpp_const.py#L166)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

BTW, I was wondering if it's safe to use double-underscore (`__`) prefix for a non-system method such as this one. (This is really a question for my Python education).
</blockquote></details>

Done. Added comments in `wrapt/wrappers.py`. Does that help?

---

*[tools/skylark/pybind.bzl, line 156 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L448nwDCmtwQyb7l39U:-L46vt7MPmwYZUNAJRuU:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c0e71feb2b272bff5400a73f04a3815430738951/tools/skylark/pybind.bzl#L156)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW typo `has`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 1 of 16 files at r3, 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[bindings/pydrake/util/cpp_const.py, line 166 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/7884#-L46BW_m2jCbGpacZRSA:-L46xHmOFxVqZ8-QCQCg:b-81nlyp) ([raw file](https://github.com/robotlocomotion/drake/blob/c0e71feb2b272bff5400a73f04a3815430738951/bindings/pydrake/util/cpp_const.py#L166)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Done. Added comments in `wrapt/wrappers.py`. Does that help?
</blockquote></details>

SGTM!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 16 files at r3, 3 of 3 files at r6.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 4 of 4 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7884)*
<!-- Sent from Reviewable.io -->
